[JSC][armv7] Use the correct calling convention for `operationWasmRethrow`

jjgriego opened the pull request:
#### 83a1bc69e51b312580fb05cb9fb683cc801ae47e
<pre>
[JSC][armv7] Use the correct calling convention for `operationWasmRethrow`
<a href="https://bugs.webkit.org/show_bug.cgi?id=250588">https://bugs.webkit.org/show_bug.cgi?id=250588</a>

Reviewed by Mark Lam.

<a href="https://github.com/WebKit/WebKit/pull/8523">https://github.com/WebKit/WebKit/pull/8523</a> changed the signature of some runtime
wasm operations, but incorrectly translated these changes for the arm32 calling
convention, breaking some tests.

Changes to other wasm operations were covered by the platform-specific
`emitCCall`, but calls to `operationWasmRethrow` are not emitted this way.

(The arm32 calling convention requires that 64-bit arguments get passed starting
in an even-numbered registers, potentially causing an odd-numbered register to
be unused in the call, in this case, `r1/argumentGPR1`)

* Source/JavaScriptCore/wasm/WasmAirIRGenerator32_64.cpp:
(JSC::Wasm::AirIRGenerator32::addRethrow):

Canonical link: <a href="https://commits.webkit.org/259152@main">https://commits.webkit.org/259152@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/fa32067b00ac1fc2464e4e62cab623344a5a5fc1

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/103951 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/13066 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/36893 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/113181 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/173485 "Built successfully and passed tests") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/14096 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/3962 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/96202 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/112273 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/109723 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/10870 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/93934 "Passed tests") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/38573 "Found 1 new test failure: imported/w3c/web-platform-tests/fetch/stale-while-revalidate/stale-image.html (failure)") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/92708 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/25545 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/80224 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/94043 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/6425 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/26926 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/90530 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/83/builds/4195 "Built successfully and passed tests") | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/6584 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/3464 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/29872 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/12576 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/46456 "Passed tests") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/99138 "Built successfully") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/6305 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/8358 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/24944 "Passed tests") | 
| | | | | 
<!--EWS-Status-Bubble-End-->

MenloDorian commented on Source/JavaScriptCore/wasm/WasmAirIRGenerator32_64.cpp:
```diff
@@ -1158,8 +1158,8 @@ auto AirIRGenerator32::addRethrow(unsigned, ControlType& data) -> PartialResult
 
     Vector<ConstrainedTmp, 3> patchArgs;
     patchArgs.append(ConstrainedTmp(instanceValue(), B3::ValueRep::reg(GPRInfo::argumentGPR0)));
-    patchArgs.append(ConstrainedTmp(TypedTmp(data.exception().lo(), Types::I32), B3::ValueRep::reg(GPRInfo::argumentGPR1)));
-    patchArgs.append(ConstrainedTmp(TypedTmp(data.exception().hi(), Types::I32), B3::ValueRep::reg(GPRInfo::argumentGPR2)));
+    patchArgs.append(ConstrainedTmp(TypedTmp(data.exception().lo(), Types::I32), B3::ValueRep::reg(GPRInfo::argumentGPR2)));
+    patchArgs.append(ConstrainedTmp(TypedTmp(data.exception().hi(), Types::I32), B3::ValueRep::reg(GPRInfo::argumentGPR3)));
```
Is this correct for all 32-bit ports?  Or is this only relevant for ARMv7?  Consider x86 and MIPS.

jjgriego replied:
It's not correct for x86 or MIPS but they shouldn't be building this file.

I can add a `#error` to flag this if someone enables this file for those platforms?

Otherwise I think I would need to rearrange things to use Air's C call support.

MenloDorian replied:
Can you upload a revised PR with the change?  It'd be good to make sure that future folks working on x86 / MIPS don't get a hard to debug issue.

MenloDorian reviewed (approved):
r=me

jjgriego committed:
[JSC][armv7] Use the correct calling convention for `operationWasmRethrow`
https://bugs.webkit.org/show_bug.cgi?id=250588

Reviewed by Mark Lam.

https://github.com/WebKit/WebKit/pull/8523 changed the signature of some runtime
wasm operations, but incorrectly translated these changes for the arm32 calling
convention, breaking some tests.

Changes to other wasm operations were covered by the platform-specific
`emitCCall`, but calls to `operationWasmRethrow` are not emitted this way.

(The arm32 calling convention requires that 64-bit arguments get passed starting
in an even-numbered registers, potentially causing an odd-numbered register to
be unused in the call, in this case, `r1/argumentGPR1`)

* Source/JavaScriptCore/wasm/WasmAirIRGenerator32_64.cpp:
(JSC::Wasm::AirIRGenerator32::addRethrow):

Canonical link: https://commits.webkit.org/259152@main
